handler/src/main/java/io/netty/handler/ssl/ResumptionController.java

```diff
@@ -0,0 +1,169 @@
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.util.internal.ObjectUtil;
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.SuppressJava6Requirement;
+
+import java.net.Socket;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Collections;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLPeerUnverifiedException;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509ExtendedTrustManager;
+
+final class ResumptionController {
+    private static final Object ENTRY = new Object();
+    private final Map<SSLEngine, Object> confirmedValidations;
+    private final AtomicReference<ResumableX509ExtendedTrustManager> resumableTm;
+
+    ResumptionController() {
+        confirmedValidations = Collections.synchronizedMap(new WeakHashMap<SSLEngine, Object>());
+        resumableTm = new AtomicReference<ResumableX509ExtendedTrustManager>();
+    }
+
+    @SuppressJava6Requirement(reason = "Guarded by version check")
+    public TrustManager wrapIfNeeded(TrustManager tm) {
+        if (tm instanceof ResumableX509ExtendedTrustManager) {
+            if (PlatformDependent.javaVersion() < 7 || !(tm instanceof X509ExtendedTrustManager)) {
+                throw new IllegalStateException(
+                        "Resumable trust manager must be a subclass of X509ExtendedTrustManager");
+            }
+            if (!resumableTm.compareAndSet(null, (ResumableX509ExtendedTrustManager) tm)) {
+                throw new IllegalStateException("Only one TrustManager can be configured for resumed sessions");
+            }
+            return new X509ExtendedWrapTrustManager((X509ExtendedTrustManager) tm, confirmedValidations);
+        }
+        return tm;
+    }
+
+    public void remove(SSLEngine engine) {
+        if (resumableTm.get() != null) {
+            confirmedValidations.remove(unwrapEngine(engine));
+        }
+    }
+
+    public boolean validateResumeIfNeeded(SSLEngine engine)
+            throws CertificateException, SSLPeerUnverifiedException {
+        ResumableX509ExtendedTrustManager tm;
+        boolean valid = engine.getSession().isValid();
+        if (valid && (tm = resumableTm.get()) != null) {
+            Certificate[] peerCertificates = engine.getSession().getPeerCertificates();
+            engine = unwrapEngine(engine);
```

Comment:
Can we add some comment to why we do the unwrapping ? I think it is because the "TrustManager" also only sees the unwrapped engine but it would be good to clarify